### PR TITLE
improve(slb_listener): improve this module and add examples

### PR DESCRIPTION
**Pull request by Lexsss**

### CODE REVIEW

File: main.tf [hunk 1]
```diff
@@ -1,48 +1,71 @@
...
+  listeners = flatten(
+    [
+      for _, obj in var.listeners : [
+        for _, id in split(",", lookup(obj, "server_group_ids")) : {
+          server_group_id = id
+          backend_port    = lookup(obj, "backend_port")
+          frontend_port   = lookup(obj, "frontend_port")
+          protocol        = lookup(obj, "protocol")
+          bandwidth       = lookup(obj, "bandwidth", "-1")
+          scheduler       = lookup(obj, "scheduler", "wrr")
```
Comment (xiaozhu36) [inline main.tf]: listener 中也应该支持获取health check，advance setting 一起其他参数中的参数，当这些参数在listener中没有指定的时候，再从health check，advance setting等中取
也就是说listener中应该支持所有的参数，但是health check中只支持health相关的，listerner中的health check只用于某个单独的listener，而health check中的health适用于所有listenre

Comment (xiaozhu36) [reply main.tf]: 其他参数也一样

**Review by xiaozhu36 — CHANGES_REQUESTED** [review]

File: README-CN.md [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+本 Module 用于在阿里云上快速创建slb listeners资源 
+
+本 Module 支持创建以下资源:
+
+* [API Slb Listener](https://www.terraform.io/docs/providers/alicloud/r/slb_listener.html)
+
+## Terraform 版本
+
+本模板要求使用版本 Terraform 0.12.
```
Comment (zql-tqy) [inline README-CN.md]: 没有加provider的版本

File: README-CN.md [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+```
+
+## 示例
+
+* [基础示例](https://github.com/terraform-alicloud-modules/terraform-alicloud-slb-listener/tree/master/examples/basic-example)
+
+## 注意事项
+
+* 本 Module 使用的 AccessKey 和 SecretKey 可以直接从 `profile` 和 `shared_credentials_file` 中获取。如果未设置，可通过下载安装 [aliyun-cli](https://github.com/aliyun/aliyun-cli#installation) 后进行配置.
```
Comment (zql-tqy) [inline README-CN.md]: 这个使用中文的标点吧

File: variables.tf [hunk 4]
```diff
@@ -3,135 +3,145 @@ variable "region" {
   default     = ""
 }
 
+variable "profile" {
+  description = "The profile name as set in the shared credentials file. If not set, it will be sourced from the ALICLOUD_PROFILE environment variable."
+  default     = ""
```
Comment (zql-tqy) [inline variables.tf:10]: 这几个变量没有type

File: variables.tf [hunk 5]
```diff
@@ -3,135 +3,145 @@ variable "region" {
...
-variable "scheduler" {
-  description = "Scheduling algorithm."
-  default     = "wrr"
-}
+
+# Deprecated variables
 
 variable "healthy_threshold" {
-  description = "Threshold determining the result of the health check is success."
+  description = "(Deprecated) Threshold determining the result of the health check is success."
```
Comment (zql-tqy) [inline variables.tf]: 这个需要写上从哪个版本开始替换，用哪个参数替换吧

File: examples/basic-example/main.tf [hunk 6]
```diff
@@ -0,0 +1,111 @@
+variable "region" {
+  default = "cn-hangzhou"
+}
+variable "zone_id" {
+  default = "cn-hangzhou-h"
+}
+
```
Comment (zql-tqy) [inline examples/basic-example/main.tf:10]: 添加上profile的参数吧

File: README-CN.md [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+## Terraform 版本
+
+本 Module 要求使用 Terraform 0.12 和 阿里云 Provider 1.56.0+。
+
+## 用法
+
+```hcl
+module "slb_listener" {
+  source  = "terraform-alicloud-modules/slb-listener/alicloud"
+  profile = "Your-Profile-Name"
```
Comment (xiaozhu36) [inline README-CN.md:20]: region

File: README-CN.md [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+    retrive_slb_ip      = "true"
+    retrive_slb_id      = "false"
+    retrive_slb_proto   = "true"
+
+    // TCP, HTTP and HTTPS
+    health_check_connect_port = "80"
+    health_check_uri          = "/"
+    health_check_http_code    = "http_2xx"
+
+    // TCP
```
Comment (xiaozhu36) [inline README-CN.md]: advance_setting 中为啥还有health check的字段

File: main.tf [hunk 9]
```diff
@@ -1,48 +1,92 @@
...
+          backend_port              = lookup(obj, "backend_port")
+          frontend_port             = lookup(obj, "frontend_port")
+          protocol                  = lookup(obj, "protocol")
+          bandwidth                 = lookup(obj, "bandwidth", "-1")
+          scheduler                 = lookup(obj, "scheduler", "wrr")
+          healthy_threshold         = lookup(obj, "healthy_threshold", null)
+          unhealthy_threshold       = lookup(obj, "unhealthy_threshold", null)
+          health_check_timeout      = lookup(obj, "health_check_timeout", null)
+          health_check_interval     = lookup(obj, "health_check_interval", null)
+          sticky_session            = lookup(obj, "sticky_session", null)
```
Comment (xiaozhu36) [inline main.tf]: 后面的null可以直接改成从health check，advance setting和x_forwarded_for中取，这样在后面具体写resource的时候就不用了

File: main.tf [hunk 10]
```diff
@@ -1,48 +1,92 @@
...
-  health_check        = var.enable_health_check ? "on" : "off"
-  gzip                = var.enable_gzip
+  sticky_session      = lookup(local.listeners[count.index], "sticky_session", lookup(var.advance_setting, "sticky_session", "off"))
+  sticky_session_type = lookup(local.listeners[count.index], "sticky_session_type", lookup(var.advance_setting, "sticky_session_type", "server"))
+  cookie              = lookup(local.listeners[count.index], "cookie", lookup(var.advance_setting, "cookie", null))
+  cookie_timeout      = lookup(local.listeners[count.index], "cookie_timeout", lookup(var.advance_setting, "cookie_timeout", "86400"))
+  health_check        = lookup(local.listeners[count.index], "health_check", lookup(var.advance_setting, "health_check", "off"))
+  gzip                = lookup(local.listeners[count.index], "gzip", lookup(var.advance_setting, "gzip", "false"))
 
   x_forwarded_for {
```
Comment (xiaozhu36) [inline main.tf:92]: x_forwarded_for 从advance setting中拿出来吧，跟 advance setting 评级，这样看的清楚些
另外，再加一个参数 ssl_certificates ，也跟 advance setting，是一个map，存放证书相关的参数, 用于https listener的设置

File: variables.tf [hunk 11]
```diff
@@ -1,137 +1,151 @@
...
+variable "listeners" {
+  description = "The list of slb listeners."
+  type        = list(map(string))
+  default     = []
 }
 
-variable "frontend_port" {
-  description = "Port used by the Server Load Balancer instance frontend."
+variable "health_check" {
+  description = "The slb listener health check settings. It supports fields 'healthy_threshold','unhealthy_threshold','health_check_timeout' and 'health_check_interval'"
```
Comment (xiaozhu36) [inline variables.tf]: 应该不止这四个参数吧，健康检查相关的都要加进来

File: examples/basic-example/main.tf [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+      port       = "80"
+      weight     = "100"
+      type       = "ecs"
+    },
+  ]
+}
+
+module "slb_listener" {
+  source  = "../../"
+  profile = var.profile
```
Comment (xiaozhu36) [inline examples/basic-example/main.tf:73]: region

File: variables.tf [hunk 13]
```diff
@@ -1,137 +1,151 @@
...
+  type        = map(number)
+  default     = {}
 }
 
-variable "bandwidth" {
-  description = "Bandwidth peak of Listener. Valid in -1/1-1000Mbps and -1 means there is no limit when SLB is 'PayByTraffic'"
-  default     = -1
+variable "advance_setting" {
+  description = "The slb listener advance settings."
```
Comment (xiaozhu36) [inline variables.tf]: advance setting中支持哪些参数，也列一下吧，多也没关系

**Review by xiaozhu36 — CHANGES_REQUESTED** [review]

File: main.tf [hunk 14]
```diff
@@ -1,48 +1,105 @@
...
+          cookie_timeout            = lookup(obj, "cookie_timeout", lookup(var.advance_setting, "cookie_timeout", "86400"))
+          gzip                      = lookup(obj, "gzip", lookup(var.advance_setting, "gzip", "false"))
+          persistence_timeout       = lookup(obj, "persistence_timeout", lookup(var.advance_setting, "persistence_timeout", null))
+          established_timeout       = lookup(obj, "established_timeout", lookup(var.advance_setting, "established_timeout", null))
+          acl_status                = lookup(obj, "acl_status", lookup(var.advance_setting, "acl_status", "off"))
+          acl_type                  = lookup(obj, "acl_type", lookup(var.advance_setting, "acl_type", null))
+          acl_id                    = lookup(obj, "acl_id", lookup(var.advance_setting, "acl_id", null))
+          idle_timeout              = lookup(obj, "idle_timeout", lookup(var.advance_setting, "idle_timeout", null))
+          request_timeout           = lookup(obj, "request_timeout", lookup(var.advance_setting, "request_timeout", null))
+          enable_http2              = lookup(obj, "enable_http2", lookup(var.advance_setting, "enable_http2", null))
```
Comment (xiaozhu36) [inline main.tf]: enable_http2 应该属于 ssl_certificates 中的参数

File: outputs.tf [hunk 15]
```diff
@@ -0,0 +1,15 @@
+// Output the ID of the new VPC and Security Group created
+
+locals {
+  this_slb_id = length(local.listeners) == 0 ? var.slb : concat(alicloud_slb_listener.this.*.load_balancer_id, [""])[0]
```
Comment (xiaozhu36) [inline outputs.tf]: 这个直接就是 this_slb_id = var.slb

File: variables.tf [hunk 16]
```diff
@@ -1,137 +1,163 @@
...
+variable "listeners" {
+  description = "The list of slb listeners, It's supports all fields of alicloud_slb_listener resource."
+  type        = list(map(string))
+  default     = []
 }
 
-variable "frontend_port" {
-  description = "Port used by the Server Load Balancer instance frontend."
+variable "health_check" {
+  description = "The slb listener health check settings. It's supports fields 'healthy_threshold','unhealthy_threshold','health_check_timeout', 'health_check', 'health_check_type', 'health_check_connect_port', 'health_check_domain', 'health_check_uri', 'health_check_http_code', 'health_check_method' and 'health_check_interval'"
```
Comment (xiaozhu36) [inline variables.tf]: description = "The slb listener health check settings to use on listeners. It's supports fields 'healthy_threshold','unhealthy_threshold','health_check_timeout', 'health_check', 'health_check_type', 'health_check_connect_port', 'health_check_domain', 'health_check_uri', 'health_check_http_code', 'health_check_method' and 'health_check_interval'"

File: variables.tf [hunk 17]
```diff
@@ -1,137 +1,163 @@
...
+  type        = map(string)
+  default     = {}
 }
 
-variable "bandwidth" {
-  description = "Bandwidth peak of Listener. Valid in -1/1-1000Mbps and -1 means there is no limit when SLB is 'PayByTraffic'"
-  default     = -1
+variable "advance_setting" {
+  description = "The slb listener advance settings, It's supports fields 'sticky_session', 'sticky_session_type', 'cookie', 'cookie_timeout', 'gzip', 'persistence_timeout', 'acl_status', 'acl_type', 'acl_id', 'idle_timeout', 'request_timeout' and 'enable_http2'"
```
Comment (xiaozhu36) [inline variables.tf]: description = "The slb listener advance settings to use on listeners. It's supports fields 'sticky_session', 'sticky_session_type', 'cookie', 'cookie_timeout', 'gzip', 'persistence_timeout', 'acl_status', 'acl_type', 'acl_id', 'idle_timeout' and 'request_timeout'."

File: variables.tf [hunk 18]
```diff
@@ -1,137 +1,163 @@
...
+  type        = map(string)
+  default     = {}
 }
 
-variable "server_group_id" {
-  description = "The ID of server group."
-  default     = ""
+variable "x_forwarded_for" {
+  description = "Whether to set additional HTTP Header field 'X-Forwarded-For' (documented below), It's supports fields 'retrive_slb_ip', 'retrive_slb_id' and 'retrive_slb_proto'"
```
Comment (xiaozhu36) [inline variables.tf]: description = "Additional HTTP Header field 'X-Forwarded-For'  to use on listeners. It's supports fields 'retrive_slb_ip', 'retrive_slb_id' and 'retrive_slb_proto'"

File: variables.tf [hunk 19]
```diff
@@ -1,137 +1,163 @@
...
+  type        = map(bool)
+  default     = {}
 }
 
-variable "scheduler" {
-  description = "Scheduling algorithm."
-  default     = "wrr"
+variable "ssl_certificates" {
+  description = "SLB Server certificate settings, It's supports fields 'tls_cipher_policy' and 'server_certificate_id'"
```
Comment (xiaozhu36) [inline variables.tf]: description = "SLB Server certificate settings to use on listeners. It's supports fields 'tls_cipher_policy', 'server_certificate_id' and 'enable_http2'"

File: variables.tf [hunk 20]
```diff
@@ -1,137 +1,163 @@
...
+}
+
+variable "skip_region_validation" {
+  description = "Skip static validation of region ID. Used by users of alternative AlibabaCloud-like APIs or users w/ access to regions that are not public (yet)."
+  type        = bool
+  default     = false
+}
 # Load Balancer Instance variables
 variable "slb" {
   description = "The load balancer ID used to add a new listener."
```
Comment (xiaozhu36) [inline variables.tf]: description = "The load balancer ID used to add one or more listeners."

File: variables.tf [hunk 21]
```diff
@@ -1,137 +1,163 @@
...
+variable "create" {
+  description = "Whether to create load balancer listeners."
+  type        = bool
+  default     = true
 }
-
-variable "backend_port" {
-  description = "Port used by the Server Load Balancer instance backend."
+variable "listeners" {
+  description = "The list of slb listeners, It's supports all fields of alicloud_slb_listener resource."
```
Comment (xiaozhu36) [inline variables.tf]: description = "List of slb listeners. Each item can set all or part fields of alicloud_slb_listener resource."

File: README.md [hunk 22]
```diff
@@ -1,43 +1,103 @@
...
+      protocol          = "tcp"
+      bandwidth         = "-1"
+      scheduler         = "wrr"
+      healthy_threshold = "4"
+      gzip              = "false"
+      health_check_type = "tcp"
+    }
+  ]
+  
+  // you can also leave this field unspecified to use the default value
```
Comment (xiaozhu36) [inline README.md]: // health_check will apply to all of listeners if health checking is not set in the listeners

File: README.md [hunk 23]
```diff
@@ -1,43 +1,103 @@
...
+    healthy_threshold         = "3"
+    unhealthy_threshold       = "2"
+    health_check_timeout      = "5"
+    health_check_interval     = "2"
+    health_check_connect_port = "80"
+    health_check_uri          = "/"
+    health_check_http_code    = "http_2xx"
+  }
+  
+  // you can also leave this field unspecified to use the default value
```
Comment (xiaozhu36) [inline README.md]: advance_setting 改为 advanced_setting
advanced_setting will apply to all of listeners if some fields are not set in the listeners

File: README.md [hunk 24]
```diff
@@ -1,43 +1,103 @@
...
+    sticky_session_type = "server"
+    cookie_timeout      = "86400"
+    gzip                = "false"
+    retrive_slb_ip      = "true"
+    retrive_slb_id      = "false"
+    retrive_slb_proto   = "true"
+    persistence_timeout = "5"
+  }
+  
+  // you can also leave this field unspecified to use the default value
```
Comment (xiaozhu36) [inline README.md]: x_forwarded_for will apply to all of listeners if it is not set in the listeners

File: README.md [hunk 25]
```diff
@@ -1,43 +1,103 @@
...
+  x_forwarded_for = {
+    retrive_slb_ip    = "true"
+    retrive_slb_id    = "false"
+    retrive_slb_proto = "true"
+  }
+  
+  // you can also leave this field unspecified to use the default value
```
Comment (xiaozhu36) [inline README.md]: // ssl_certificates will apply to all of https listeners if it is not set in the listeners

**Review by xiaozhu36 — CHANGES_REQUESTED** [review]

**Review by xiaozhu36 — APPROVED** [review]

**Pull request merged**